src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverImpl.java

```diff
@@ -1054,14 +1058,41 @@ public boolean hasChildren(Resource resource) {
     public String getParentResourceType(final Resource resource) {
         String resourceSuperType = null;
         if ( resource != null ) {
-            resourceSuperType = resource.getResourceSuperType();
-            if (resourceSuperType == null) {
-                resourceSuperType = this.getParentResourceType(resource.getResourceType());
+            if (getParentResourceTypeMap().containsKey(resource.getPath())) {
+                resourceSuperType = getParentResourceTypeMap().get(resource.getPath());
+            } else {
+                resourceSuperType = getParentResourceTypeInternal(resource);
+                getParentResourceTypeMap().put(resource.getPath(), resourceSuperType);
             }
         }
         return resourceSuperType;
     }
 
+    String getParentResourceTypeInternal(final Resource resource) {
+        String resourceSuperType;
+        resourceSuperType = resource.getResourceSuperType();
+        if (resourceSuperType == null) {
+            resourceSuperType = this.getParentResourceType(resource.getResourceType());
+        }
+        return resourceSuperType;
+    }
+    
+    /**
+     * get the map to hold the resourceType - parentResourceType relations
+     * @return the map
+     */
+    private @NotNull Map<String,String> getParentResourceTypeMap() {
+        @SuppressWarnings("unchecked")
+        Map<String,String> result = (Map<String, String>) getPropertyMap().get(PARENT_RT_CACHEKEY);
+        if (result == null) {
+            result = new HashMap<>();
+            getPropertyMap().put(PARENT_RT_CACHEKEY, result);
+        }
+        return result;
+    }
```

Comment:
This can be simplified:

```suggestion
    private @NotNull Map<String,String> getParentResourceTypeMap() {
        @SuppressWarnings("unchecked")
        return (Map<String, String>) getPropertyMap().computeIfAbsent(PARENT_RT_CACHEKEY, k -> new HashMap<>());
    }
```